# Welcome to keio tutorial !!!
These tutorials walk you through the process of analyzing illumina read for developing keio collection type results in our model organism. 

![alt text](https://github.com/ravinpoudel/keio/blob/master/keio.png)


The Keio collection in  Escherichia coli K-12 represents  a collection of single-gene deleted mutants. 
This collection was created by manually one by one by replacing predicted ORF with a kanamycin cassette to inactivate chromosomal genes. Then primers were designed to create in-frame deletions upon excision of the resistance cassette. Of 4288 genes targeted, 3985 mutants were obtained and major of these mutants represents the mutation of non-essential genes.
Primarily, the KEIO collection provides a new molecular tool/resource to understand the functional and physiological aspects of gene at the system levels. 

Although, creating such molecular collection / tools takes lot of resorce and daunting. Thus, here we explore and RB-TnSeq (Randomly Barcode Transposons) method to crate a single gene mutant type collection in Phaeobacter_inhibens_DSM_17395. The molecular construct of Randomly Barcode Transposons is similar follow:

![alt text](https://github.com/ravinpoudel/keio/blob/master/KEIO_mutant_summary.png)


RbTransposon.png

RB_Clone.png
